packages/core/src/components/forms/numericInput.tsx

```diff
@@ -351,11 +375,33 @@ export class NumericInput extends AbstractComponent<HTMLInputProps & INumericInp
             this.incrementValue(delta);
         }
 
+        if (this.props.allowNumericCharactersOnly && this.isKeyboardEventDisabledForBasicNumericEntry(e)) {
+            e.preventDefault();
+        }
+
         Utils.safeInvoke(this.props.onKeyDown, e);
     }
 
+    private handleInputPaste = (e: React.ClipboardEvent<HTMLInputElement>) => {
+        this.didPasteEventJustOccur = true;
+        Utils.safeInvoke(this.props.onPaste, e);
+    }
+
     private handleInputChange = (e: React.FormEvent<HTMLInputElement>) => {
-        const nextValue = (e.target as HTMLInputElement).value;
+        const value = (e.target as HTMLInputElement).value;
+
+        let nextValue: string;
+
+        if (this.props.allowNumericCharactersOnly && this.didPasteEventJustOccur) {
+            this.didPasteEventJustOccur = false;
```

Comment:
It's a little extra work, yes—with my interviewer hat on, _O(n)_ for each keystroke—but that won't be an issue in real-world scenarios where numbers are only going to be a few characters long. I'd still prefer not to do that extra work, if only because it feels unclean and less clear that it's helpful specifically for the paste scenario without an additional comment.